svcdef/svcdef.go

```diff
@@ -0,0 +1,347 @@
+	// this FieldType represents a Map. If not, Map is nil.
+	Map *Map
+	// StarExpr is True if this FieldType represents a pointer to a type.
+	StarExpr bool
+	// ArrayType is True if this FieldType represents a slice of a type.
+	ArrayType bool
+}
+
+// HTTPBinding represents one of potentially several bindings from a gRPC
+// service method to a particuar HTTP path/verb.
+type HTTPBinding struct {
+	Verb string
+	Path string
+	// There is one HTTPParamter for each of the parent service methods Fields.
+	Params []*HTTPParameter
+}
+
+// HTTPParameter represents the location of one field for a given HTTPBinding.
+// Each HTTPParameter corresponds to one Field of the parent
+// ServiceMethod.RequestType.Fields
+type HTTPParameter struct {
+	// Field points to a Field on the Parent service methods "RequestType".
+	Field *Field
+	// Location will be either "body", "path", or "query"
+	Location string
+}
+
+func retrieveTypeSpecs(f *ast.File) ([]*ast.TypeSpec, error) {
+	var rv []*ast.TypeSpec
+	for _, dec := range f.Decls {
+		switch gendec := dec.(type) {
+		case *ast.GenDecl:
+			for _, spec := range gendec.Specs {
+				switch ts := spec.(type) {
+				case *ast.TypeSpec:
+					rv = append(rv, ts)
+				}
+			}
+		}
+	}
+	return rv, nil
+}
+
+func New(goFiles []io.Reader, protoFiles []io.Reader) (*Svcdef, error) {
+	rv := Svcdef{}
+
+	for _, gofile := range goFiles {
+		fset := token.NewFileSet()
+		fileAst, err := parser.ParseFile(fset, "", gofile, parser.ParseComments)
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't parse go file to create Svcdef")
+		}
+
+		typespecs, _ := retrieveTypeSpecs(fileAst)
+		for _, t := range typespecs {
+			//sp.Dump(t)
+			switch typdf := t.Type.(type) {
+			case *ast.Ident:
+				if typdf.Name == "int32" {
+					nenm, err := NewEnum(t)
+					if err != nil {
+						return nil, errors.Wrapf(err, "error parsing enum %q", t.Name.Name)
+					}
+					rv.Enums = append(rv.Enums, nenm)
+				}
+			case *ast.StructType:
+				nmsg, err := NewMessage(t)
+				if err != nil {
+					return nil, errors.Wrapf(err, "error parsing message %q", t.Name.Name)
+				}
+				rv.Messages = append(rv.Messages, nmsg)
+			case *ast.InterfaceType:
+				// Each service will have two interfaces ("{SVCNAME}Server" and
+				// "{SVCNAME}Client") each containing the same information that we
+				// care about, but structured a bit differently. For simplicity,
+				// skip the "Client" interface.
+				if strings.HasSuffix("Client", t.Name.Name) {
+					break
+				}
+				nsvc, err := NewService(t)
+				if err != nil {
+					return nil, errors.Wrapf(err, "error parsing service %q", t.Name.Name)
+				}
+				rv.Service = nsvc
+			}
+		}
+	}
+	resolveTypes(&rv)
+	err := ConsolidateHTTP(&rv, protoFiles)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to consolidate HTTP")
+	}
+
+	return &rv, nil
+}
+
+// NewEnum returns a new Enum struct derived from an *ast.TypeSpec
```

Comment:
This comment is the same as reading the function signature. We should avoid having these types of comments unless they say something that is non obvious from the function signature as the documentation contains these. 

![2016-10-13-142056_562x148_scrot](https://cloud.githubusercontent.com/assets/19537819/19367734/3ea02964-9151-11e6-84a5-2de8e054678c.png)